파이썬 머신러닝 완벽 가이드 개정 2판 8장 p509~534

# 8. 텍스트 분석

## 04. 텍스트 분류 실습 - 20 뉴스그룹 분류

텍스트 분류 : 특정 문서의 분류를 학습 데이터를 통해 학습해 모델을 생성한 뒤, 이 학습 모델을 이용해 다른 문서의 분류를 예측하는 것

**사이킷런의 fetch_20newsgroups() API**

 : 뉴스그룹의 분류를 수행해 볼 수 있는 예제 데이터를 제공

 - 텍스트를 피처 벡터화로 변환 -> 희소 행렬 형태가 됨
 - 희소 행렬에 분류를 효과적으로 처리하는 알고리즘 : 로지스틱 회귀, 선형 서포트 벡터 머신, 나이브 베이즈 등
 - 로지스틱 회귀 사용해보기






### [텍스트 정규화]

In [13]:
from sklearn.datasets import fetch_20newsgroups

news_data = fetch_20newsgroups(subset='all', random_state=156)

In [14]:
print(news_data.keys())

dict_keys(['data', 'filenames', 'target_names', 'target', 'DESCR'])


In [15]:
import pandas as pd
print('target 클래스의 값과 분포도 \n', pd.Series(news_data.target).value_counts().sort_index()) # 0~19, 20개
print('target 클래스의 이름들 \n', news_data.target_names)

target 클래스의 값과 분포도 
 0     799
1     973
2     985
3     982
4     963
5     988
6     975
7     990
8     996
9     994
10    999
11    991
12    984
13    990
14    987
15    997
16    910
17    940
18    775
19    628
Name: count, dtype: int64
target 클래스의 이름들 
 ['alt.atheism', 'comp.graphics', 'comp.os.ms-windows.misc', 'comp.sys.ibm.pc.hardware', 'comp.sys.mac.hardware', 'comp.windows.x', 'misc.forsale', 'rec.autos', 'rec.motorcycles', 'rec.sport.baseball', 'rec.sport.hockey', 'sci.crypt', 'sci.electronics', 'sci.med', 'sci.space', 'soc.religion.christian', 'talk.politics.guns', 'talk.politics.mideast', 'talk.politics.misc', 'talk.religion.misc']


-> Target 클래스의 값 : 0부터 19까지 20개로 구성

In [16]:
print(news_data.data[0])

From: egreen@east.sun.com (Ed Green - Pixel Cruncher)
Subject: Re: Observation re: helmets
Organization: Sun Microsystems, RTP, NC
Lines: 21
Distribution: world
Reply-To: egreen@east.sun.com
NNTP-Posting-Host: laser.east.sun.com

In article 211353@mavenry.altcit.eskimo.com, maven@mavenry.altcit.eskimo.com (Norman Hamer) writes:
> 
> The question for the day is re: passenger helmets, if you don't know for 
>certain who's gonna ride with you (like say you meet them at a .... church 
>meeting, yeah, that's the ticket)... What are some guidelines? Should I just 
>pick up another shoei in my size to have a backup helmet (XL), or should I 
>maybe get an inexpensive one of a smaller size to accomodate my likely 
>passenger? 

If your primary concern is protecting the passenger in the event of a
crash, have him or her fitted for a helmet that is their size.  If your
primary concern is complying with stupid helmet laws, carry a real big
spare (you can put a big or small head in a big helmet, bu

뉴스 그룹 기사의 내용, 뉴스그룹 제목, 작성자, 소속, 이메일 등 다양한 정보를 가지고 있음
- 구성된 기사 내용으로 어떤 뉴스 그룹에 속하는지 분류하기 위해 다른 정보는 제거하기

fetch_20newsgroups()는 subset 파라미터를 이용해 학습 데이터 세트와 테스트 데이터 세트를 분리해 내려받을 수 있음

In [17]:
from sklearn.datasets import fetch_20newsgroups
# subset='train'으로 학습용 데이터만 추출, remove=('headers', 'footers', 'quotes')로 내용만 추출
train_news = fetch_20newsgroups(subset = 'train', remove = ('headers','footers','quotes'),random_state = 156)

X_train = train_news.data
y_train = train_news.target

# subset='test'으로 테스트 데이터만 추출, remove=('headers', 'footers', 'quotes')로 내용만 추출
test_news= fetch_20newsgroups(subset = 'test', remove = ('headers','footers','quotes'), random_state = 156)

X_test = test_news.data
y_test = test_news.target

print('학습 데이터 크기 {0}, 테스트 데이터 크기 {1}'.format(len(train_news.data),len(test_news.data)))


학습 데이터 크기 11314, 테스트 데이터 크기 7532


### [피처 벡터화 변환과 머신러닝 모델 학습/예측/평가]

- CountVectorizer를 이용해 학습 데이터의 텍스트를 피처 벡터화하기
- 테스트 데이터를  CountVectorizer를 이용해 피처 벡터화할 때 :
 - 학습 데이터를 이용해 fit()이 수행된 CountVectorizer 객체를 이용해 테스트 데이터를 변환해야 함
 - 그래야만 학습 시 설정된 CountVectorizer의 피처 개수와 테스트 데이터를  CountVectorizer로 변환할 피처 개수가 같아짐

In [18]:
from sklearn.feature_extraction.text import CountVectorizer

# Count Vectorization으로 피처 벡터화 변환 수행.
cnt_vect = CountVectorizer()
cnt_vect.fit(X_train)
X_train_cnt_vect = cnt_vect.transform(X_train)

# 학습 데이터로 fit( )된 CountVectorizer를 이용해 테스트 데이터를 피터 벡터화 변환 수행.
X_test_cnt_vect = cnt_vect.transform(X_test) ## 다시 fit을 하거나 fit_transform()을 사용하면 xxxx

print('학습 데이터 텍스트의 CountVectorizer Shape:', X_train_cnt_vect.shape)

학습 데이터 텍스트의 CountVectorizer Shape: (11314, 101631)


-> 11315개의 문서에서 101631개의 피처(단어)가 생성됨

피처 벡터화된 데이터에 로지스틱 회귀를 적용해 뉴스그룹에 대한 분류를 예측해보기

In [19]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import warnings
warnings.filterwarnings('ignore')

# LogisticRegression을 이용하여 학습/예측/평가 수행.
lr_clf = LogisticRegression(solver='liblinear')
lr_clf.fit(X_train_cnt_vect,y_train)
pred = lr_clf.predict(X_test_cnt_vect)
print('CountVectorized Logistic Regression의 예측 정확도는 {0: .3f}',format(accuracy_score(y_test,pred)))

CountVectorized Logistic Regression의 예측 정확도는 {0: .3f} 0.6168348380244291


-> Count 기반으로 피처 벡터화가 적용된 데이터 세트에 대한 로지스틱 회귀의 예측 정확도

Count 기반에서 TF_IDF 기반으로 벡터화를 변경해 예측 모델을 수행

In [20]:
from sklearn.feature_extraction.text import TfidfVectorizer

# TF-IDF 벡터화를 적용해 학습 데이터 세트와 테스트 데이터 세트 변환.
tfidf_vect = TfidfVectorizer()
tfidf_vect.fit(X_train)
X_train_tfidf_vect = tfidf_vect.transform(X_train)
X_test_tfidf_vect = tfidf_vect.transform(X_test)

# LogisticRegression을 이용해 학습/예측/평가 수행.
lr_clf = LogisticRegression(solver='liblinear')
lr_clf.fit(X_train_tfidf_vect, y_train)
pred = lr_clf.predict(X_test_tfidf_vect)
print('TF-IDF Logistic Regression의 예측 정확도는 {0:.3f}'.format(accuracy_score(y_test, pred)))

TF-IDF Logistic Regression의 예측 정확도는 0.678


-> TF-IDF가 단순 카운트 기반보다 훨씬 높은 예측 정확도를 제공

더 다양한 파라미터를 적용해보기

In [21]:
# stop words 피처링을 추가하고 ngram을 기본 (1,1)에서 (1,2)로 변경해 피처 벡터화 적용.
tfidf_vect = TfidfVectorizer(stop_words='english',ngram_range=(1,2),max_df = 300)
tfidf_vect.fit(X_train)
X_train_tfidf_vect = tfidf_vect.transform(X_train)
X_test_tfidf_vect = tfidf_vect.transform(X_test)

lf_clf = LogisticRegression(solver='liblinear')
lf_clf.fit(X_train_tfidf_vect,y_train)
lf_clf.predict(X_test_tfidf_vect)
print('TF-IDF Vectorized Logistic Regression 의 예측 정확도는 {0:.3f}'.format(accuracy_score(y_test, pred)))


TF-IDF Vectorized Logistic Regression 의 예측 정확도는 0.678


GridSearchCV를 이용해 로지스틱 회귀의 하이퍼파라미터 최적화 수행하기
- 로지스틱 회귀의 C 파라미터만 변경하면서 최적의 C값을 찾은 뒤, 이 C 값으로 학습된 모델에서 테스트 데이터로 예측해 성능 평가하기

In [22]:
from sklearn.model_selection import GridSearchCV

# 최적 C값 도출 튜닝 수행. CV는 3폴드 세트로 설정.
params = {'C':[0.01,0.1,1,5,10]} # C: 규제(strength of regularization)의 강도를 조절하는 하이퍼파라미터
grid_cv_lr = GridSearchCV(lf_clf, param_grid = params, cv = 3, scoring = 'accuracy', verbose = 1)
grid_cv_lr.fit(X_train_tfidf_vect, y_train)
print('Logistic Regression best C parameter :', grid_cv_lr.best_params_ )

# 최적 C값으로 학습된 grid_cv로 예측 및 정확도 평가.
pred = grid_cv_lr.predict(X_test_tfidf_vect)
print('TF-IDF Vectorized Logistic Regression의 예측 정확도는 {0:.3f}'.format(accuracy_score(y_test, pred)))

Fitting 3 folds for each of 5 candidates, totalling 15 fits
Logistic Regression best C parameter : {'C': 10}
TF-IDF Vectorized Logistic Regression의 예측 정확도는 0.704


### [사이킷런 파이프라인 사용 및 GridSearchCV와의 결합]

사이킷런 파이프라인 클래스를 이용 -> 피처 벡터화와 ML 알고리즘 학습/예측을 위한 코드 작성을 한 번에 진행할 수 있음
- 머신러닝에서 Pipeline을 이용해 데이터의 전처리와 머신러닝 학습 과정을 통일한 API 기반에서 처리 -> 직관적인 ML 모델 코드 생성
- 대용량 데이터의 피처 벡터화 결과를 별도 데이터로 저장 x, 스트림 기반에서 바로 머신러닝 알고리즘의 데이터로 입력할 수 있음 -> 수행 시간 절약

Pipeline 객체를 다음과 같이 선언

In [25]:
from sklearn.pipeline import Pipeline

pipeline = Pipeline([('tfidf_vect',TfidfVectorizer(stop_words = 'english')),
                     ('lr_clf',LogisticRegression(random_state = 156))])

- TfidfVectorizer 객체를 tfidf_vect라는 객체 변수명으로, LogisticRegression 객체를 lr_clf 라는 객체 변수명으로 생성한 뒤, 이 두 개의 객체를 파이프라인으로 연결하는 PIpeline 객체 pipeline

In [26]:
from sklearn.pipeline import Pipeline

# TfidfVectorizer 객체를 tfidf_vect로, LogisticRegression 객체를 lr_clf로 생성하는 Pipeline 생성
pipeline = Pipeline([('tfidf_vect',TfidfVectorizer(stop_words = 'english',ngram_range=(1,2),max_df = 300)),
                     ('lr_clf',LogisticRegression(solver = 'liblinear', C = 10))])

# 별도의 TfidfVectorizer 객체의 fit( ), transform( )과 LogisticRegression의 fit(), predict( )가 필요 없음.
# pipeline의 fit( )과 predict( )만으로 한꺼번에 피처 벡터화와 ML 학습/예측이 가능.
pipeline.fit(X_train,y_train)
pred = pipeline.predict(X_test)
print('Pipeline을 통한 Logistic Regression의 예측 정확도는 {0:.3f}'.format(accuracy_score(y_test, pred)))

Pipeline을 통한 Logistic Regression의 예측 정확도는 0.704


GridSearchCV에 Pipeline을 입력하면서 TfidfVectorizer 의 파라미터와 LogisticRegression 의 하이퍼 파라미터를 함께 최적화함

Pipeline + GridSearchCV
- GridSearchCV에 Estimator가 아닌 Pipeline을 입력 -> param_grid 의 입력값 설정이 기존과 다름
 - 딕셔너리 형태의 Key와 Value 값을 가짐
 - Value를 리스트 형태로 입력한는 것은 동일

In [ ]:
#from sklearn.pipeline import Pipeline

#pipeline = Pipeline([
    #('tfidf_vect', TfidfVectorizer(stop_words = 'english')),
    #('lr_clf', LogisticRegression())
#])

# Pipeline에 기술된 각각의 객체 변수에 언더바(_) 2개를 연달아 붙여 GridSearchCV에 사용될
# 파라미터/하이퍼 파라미터 이름과 값을 설정.
#params = {'tfidf_vect__ngram_range':[(1,1),(1,2),(1,3)],
          #'tfidf_vect__max_df':[100,300,700],
          #'lr_clf__C': [1,5,10]

#}

# GridSearchCV의 생성자에 Estimator가 아닌 Pipeline 객체 입력
#grid_cv_pipe = GridSearchCV(pipeline, param_grid = params, cv = 3, scoring = 'accuracy', verbose = 1)
#grid_cv_pipe.fit(X_train,y_train)
#print(grid_cv_pipe.best_params_, grid_cv_pipe.best_score_)

#pred = grid_cv_pipe.predict(X_test)
#print('Pipeline을 통한 Logistic Regression의 예측 정확도는 {0:.3f}'.format(accuracy_score(y_test, pred)))

## 05. 감성분석
### [감성 분석 소개]
감성 분석
- 문서의 주관적인 감성/의견/감정/기분 등을 파악하기 위한 방법
- 소셜 미디어, 여론조사, 온라인 리뷰, 피드백 등 다양한 분야에서 활용
- 문서 내 텍스트가 나타내는 여러 가지 주고나적인 단어와 문맥을 기반으로 감성 수치를 계산하는 방법을 이용

감성분석의 지도학습/비지도학습
- 지도학습
 - 학습 데이터와 타깃 레이블 값을 기반으로 감성 분석 학습을 수행
 - 다른 데이터의 감성 분석을 예측하는 방법
 - 일반적인 텍스트 기반의 분류와 거의 동일
- 비지도학습
 - 'Lexicon' 이라는 일종의 감성 어휘 사전을 이용
 - Lexicon : 감성 분석을 위한 용어와 문맥에 대한 다양한 정보, 이를 통해 문서의 긍정/부정적 감성 여부 판단

### [지도학습 기반 감성 분석 실습 - IMDB 영화평]
IMDB의 영화 사이트의 영화평을 이용
- 영화평의 텍스트를 분석해 감성 분석 결과가 긍정/부정인지 예측

In [27]:
import pandas as pd

review_df = pd.read_csv('/content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/labeledTrainData.tsv',
                         header=0, sep='\t', quoting=3)
review_df.head(3)

id  sentiment                                             review
0  "5814_8"          1  "With all this stuff going down at the moment ...
1  "2381_9"          1  "\"The Classic War of the Worlds\" by Timothy ...
2  "7759_3"          0  "The film starts with a manager (Nicholas Bell...

- id : 각 데이터의 id
- sentiment : 영화평(review)의 Sentiment 결과 값 (Target Label), 1은 긍정적 평가, 0은 부정적 평가
- review : 영화평의 텍스트

In [28]:
print(review_df['review'][0])

"With all this stuff going down at the moment with MJ i've started listening to his music, watching the odd documentary here and there, watched The Wiz and watched Moonwalker again. Maybe i just want to get a certain insight into this guy who i thought was really cool in the eighties just to maybe make up my mind whether he is guilty or innocent. Moonwalker is part biography, part feature film which i remember going to see at the cinema when it was originally released. Some of it has subtle messages about MJ's feeling towards the press and also the obvious message of drugs are bad m'kay.<br /><br />Visually impressive but of course this is all about Michael Jackson so unless you remotely like MJ in anyway then you are going to hate this and find it boring. Some may call MJ an egotist for consenting to the making of this movie BUT MJ and most of his fans would say that he made it for the fans which if true is really nice of him.<br /><br />The actual feature film bit when it finally sta

- replace()를 str에 적용해 'br/' 태그를 공백으로 모두 바꾸기
- 영어가 아닌 숫자/특수문자도 별 의미가 없으므로 공란으로 변경

In [29]:
import re

# <br> html 태그는 replace 함수로 공백으로 변환
review_df['review'] = review_df['review'].str.replace('<br />', ' ')

# 파이썬의 정규 표현식 모듈인 re를 이용해 영어 문자열이 아닌 문자는 모두 공백으로 변환
review_df['review'] = review_df['review'].apply(lambda x : re.sub('[^a-zA-Z]', ' ' , x))

- 결정 값 클래스인 sentiment 칼럼을 별도로 추출해 결정 값 데이터 세트 만들기
- 원본 데이터 세트에서 id와 sentiment 칼럼을 삭제해 피처 데이터 세트를 생성

In [30]:
from sklearn.model_selection import train_test_split

class_df = review_df['sentiment']
feature_df = review_df.drop(['id', 'sentiment'], axis=1, inplace=False)

X_train, X_test, y_train, y_test = train_test_split(feature_df , class_df, test_size=0.3, random_state=156)
X_train.shape, X_test.shape

((17500, 1), (7500, 1))

-> 학습용 데이터는 17500개의 리뷰, 테스트용 데이터는 7500개의 리뷰

감상평(Review) 텍스트를 피처 벡터화한 후 ML 분류 알고리즘을 적용해 예측 성능을 측정하기
- Count 벡터화를 적용해 예측 성능을 측정
- TF-IDF 벡터화를 적용
- 예측 성능 평가 : 테스트 데이터 세트의 정확도와 ROD-AUC를 모두 측정

In [31]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, roc_auc_score

# 스톱 워드는 English, ngram은 (1, 2)로 설정해 CountVectorization 수행.
# LogisticRegression의 C는 10으로 설정.
pipeline = Pipeline([
    ('cnt_vect', CountVectorizer(stop_words='english', ngram_range=(1, 2))),
    ('lr_clf', LogisticRegression(solver='liblinear', C=10))])

# Pipeline 객체를 이용해 fit(), predict()로 학습/예측 수행. predict_proba()는 roc_auc 때문에 수행.
pipeline.fit(X_train['review'], y_train)
pred = pipeline.predict(X_test['review'])
pred_probs = pipeline.predict_proba(X_test['review'])[:,1]

print('예측 정확도는 {0:.4f}, ROC-AUC는 {1:.4f}'.format(accuracy_score(y_test, pred), roc_auc_score(y_test, pred_probs)))


예측 정확도는 0.8861, ROC-AUC는 0.9503


TF-IDF 벡터화를 적용해 다시 예측 성능을 측정하기
- Pipeline에서 CountVectorizer를 TfidfVectorizer로 변경

In [32]:
# 스톱 워드는 english, filtering, ngram은 (1, 2)로 설정해 TF-IDF 벡터화 수행.
# LogisticRegression의 C는 10으로 설정.
pipeline = Pipeline([
    ('tfidf_vect', TfidfVectorizer(stop_words='english', ngram_range=(1, 2))),
    ('lr_clf', LogisticRegression(solver='liblinear', C=10))])

pipeline.fit(X_train['review'], y_train)
pred = pipeline.predict(X_test['review'])
pred_probs = pipeline.predict_proba(X_test['review'])[:,1]
print('예측 정확도는 {0:.4f}, ROC-AUC는 {1:.4f}'.format(accuracy_score(y_test, pred), roc_auc_score(y_test, pred_probs)))


예측 정확도는 0.8936, ROC-AUC는 0.9598


-> TF-IDF 기반 피처 벡터화의 예측 성능이 조금 더 나아짐

### [비지도학습 기반 감성 분석 소개]

비지도 감성 분석 : Lexicon을 기반으로 함
- 많은 감성 분석용 데이터는 결정된 레이블 값을 가지지 않음 -> Lexicon을 유용하게 사용

(Lexicon을 감성 사전으로 표현하기)
- 감성 지수 : 긍정/부정 감성의 정도를 의미하는 수치를 가짐
- 감성 지수는 단어의 위치, 주변 단어, 문맥, POS(part of speech) 등을 참고해 결정
- NLTK 패키지

NLTK 패키지
- 시맨틱 : 문맥상 의미, 동일한 단어나 문장이라도 다른 환경과 문맥에선느 다르게 표현되거나 이해될 수 있음
- WordNet : 방대한 영어 어휘 사전, 시맨틱 분석을 제공하는 어휘 사전
- WordNet은 다양한 상황에서 같은 어휘라도 다르게 사용하는 어휘의 시맨틱 정보를 제공하며, 각각의 품사로 구성된 개별 단어를 Synset이라는 개념을 이용해 표현함

감성 사전
- NLTK
- SentiWordNet
- VADER
- Pattern


### [SentiWordNet을 이용한 감성 분석]
#### WordNet Synset과 SentiWordNet SentiSynset 클래스의 이해


In [33]:
import nltk
nltk.download('all')

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/abc.zip.
[nltk_data]    | Downloading package alpino to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/alpino.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping taggers/averaged_perceptron_tagger.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger_eng to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping
[nltk_data]    |       taggers/averaged_perceptron_tagger_eng.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger_ru to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping
[nltk_data]    |       taggers/averaged_perceptron_tagger_ru.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger_rus to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |  

True

In [34]:
from nltk.corpus import wordnet as wn

term = 'present'

# 'present'라는 단어로 wordnet의 synsets 생성
synsets = wn.synsets(term)
print('synsets() 반환 type :', type(synsets))
print('synsets() 반환 값 개수 :', len(synsets))
print('synsets() 반환 값 :', synsets)

synsets() 반환 type : <class 'list'>
synsets() 반환 값 개수 : 18
synsets() 반환 값 : [Synset('present.n.01'), Synset('present.n.02'), Synset('present.n.03'), Synset('show.v.01'), Synset('present.v.02'), Synset('stage.v.01'), Synset('present.v.04'), Synset('present.v.05'), Synset('award.v.01'), Synset('give.v.08'), Synset('deliver.v.01'), Synset('introduce.v.01'), Synset('portray.v.04'), Synset('confront.v.03'), Synset('present.v.12'), Synset('salute.v.06'), Synset('present.a.01'), Synset('present.a.02')]


-> 총 18개의 서로 다른 semantic을 가지는 synset 객체가 반환됨

synset 객체가 가지는 여러 가지 속성을 살핌
- POS, 정의, 부명제 등으로 시맨틱적인 요소를 표현할 수 있음

In [35]:
for synset in synsets :
    print('##### Synset name :', synset.name(), '#####')
    print('POS:', synset.lexname())
    print('Definition:', synset.definition())
    print('Lemmas:', synset.lemma_names())

##### Synset name : present.n.01 #####
POS: noun.time
Definition: the period of time that is happening now; any continuous stretch of time including the moment of speech
Lemmas: ['present', 'nowadays']
##### Synset name : present.n.02 #####
POS: noun.possession
Definition: something presented as a gift
Lemmas: ['present']
##### Synset name : present.n.03 #####
POS: noun.communication
Definition: a verb tense that expresses actions or states at the time of speaking
Lemmas: ['present', 'present_tense']
##### Synset name : show.v.01 #####
POS: verb.perception
Definition: give an exhibition of to an interested audience
Lemmas: ['show', 'demo', 'exhibit', 'present', 'demonstrate']
##### Synset name : present.v.02 #####
POS: verb.communication
Definition: bring forward and present to the mind
Lemmas: ['present', 'represent', 'lay_out']
##### Synset name : stage.v.01 #####
POS: verb.creation
Definition: perform (a play), especially on a stage
Lemmas: ['stage', 'present', 'represent']
##### Sy

-> Synset('present.n.01') 과 Synset('present.n.02')는 명사지만 서로 다은 의미를 가짐

path_similarity() : WordNet은 어떤 어휘와 다른 어휘 간의 관계를 유사도로 나타냄


In [37]:
# synset 객체를 단어별로 생성
tree = wn.synset('tree.n.01')
lion = wn.synset('lion.n.01')
tiger = wn.synset('tiger.n.02')
cat = wn.synset('cat.n.01')
dog = wn.synset('dog.n.01')

entities = [tree, lion, tiger, cat, dog]
similarities = []
entity_names = [entity.name().split('.')[0] for entity in entities]

# 단어별 synset을 반복하면서 다른 단어의 synset과 유사도를 측정합니다
for entity in entities:
  similarity = [round(entity.path_similarity(compared_entity),2) for compared_entity in entities]
  similarities.append(similarity)

# 개별 단어별 synset과 다른 단어의 synset과의 유사도를 DataFrame 형태로 저장합니다.
similarity_df = pd.DataFrame(similarities, columns = entity_names, index=entity_names)
similarity_df

tree  lion  tiger   cat   dog
tree   1.00  0.07   0.07  0.08  0.12
lion   0.07  1.00   0.33  0.25  0.17
tiger  0.07  0.33   1.00  0.25  0.17
cat    0.08  0.25   0.25  1.00  0.20
dog    0.12  0.17   0.17  0.20  1.00

In [38]:
import nltk
from nltk.corpus import sentiwordnet as swn

senti_synsets = list(swn.senti_synsets('slow'))
print('senti_synsets() 반환 type :', type(senti_synsets))
print('senti_synsets() 반환 값 개수 :', len(senti_synsets))
print('senti_synsets() 반환 값 :', senti_synsets)

senti_synsets() 반환 type : <class 'list'>
senti_synsets() 반환 값 개수 : 11
senti_synsets() 반환 값 : [SentiSynset('decelerate.v.01'), SentiSynset('slow.v.02'), SentiSynset('slow.v.03'), SentiSynset('slow.a.01'), SentiSynset('slow.a.02'), SentiSynset('dense.s.04'), SentiSynset('slow.a.04'), SentiSynset('boring.s.01'), SentiSynset('dull.s.08'), SentiSynset('slowly.r.01'), SentiSynset('behind.r.03')]


SentiSynset 객체
- 감성 지수 : 단어의 감성을 나타냄
 - 긍정 감성 지수
 - 부정 감성 지수
- 객관성 지수 : 객관성을 나타냄

어떤 단어가 전혀 감성적이지 않으면
- 객관성 지수 : 1
- 감성 지수 : 0




In [39]:
import nltk
from nltk.corpus import sentiwordnet as swn

father = swn.senti_synset('father.n.01')

print('father 긍정감성 지수:', father.pos_score())
print('father 부정감성 지수:', father.neg_score())
print('father 객관성 지수:', father.obj_score())
print('\n')
fabulous = swn.senti_synset('fabulous.a.01')
print('fabulous 긍정감성 지수:', fabulous.pos_score())
print('fabulous 부정감성 지수:', fabulous.neg_score())

father 긍정감성 지수: 0.0
father 부정감성 지수: 0.0
father 객관성 지수: 1.0


fabulous 긍정감성 지수: 0.875
fabulous 부정감성 지수: 0.125


#### SentiWordNet을 이용한 영화 감상평 감성 분석
IMDB 영화 감상평 감성 분석을 SentiWordNet Lexicon 기반으로 수행
1. 문서를 문장 단위로 분해
2. 다시 문장을 단어 단위로 토큰화하고 품사 태깅
3. 품사 태깅된 단어 기반으로 synset 객체와 senti_synset 객체를 생성
4. senti_synset 객체에서 긍정 감성/부정 감성 지수를 구하고 이를 모두 합산해 특정 임게치 값 이상일 때 긍정 감성으로, 그렇지 않을 때 부정 감성으로 결정

품사 태깅을 수행하는 내부 함수 생성

In [41]:
from nltk.corpus import wordnet as wn

# 간단한 NTLK PennTreebank Tag를 기반으로 WordNet기반의 품사 Tag로 변환
def penn_to_wn(tag):
    if tag.startswith('J'):
        return wn.ADJ
    elif tag.startswith('N'):
        return wn.NOUN
    elif tag.startswith('R'):
        return wn.ADV
    elif tag.startswith('V'):
        return wn.VERB
    return

문서를 문장 -> 단어 토큰 -> 품사 태깅 후에 SentiSynset 클래스를 생성하고 Polarity Score를 합산하는 함수 생성
- 각 단어의 긍정 감성 지수와 부정 감성 지수를 모두 합한 총 감성 지수가 0 이상일 때 긍정 감성, 그렇지 않으면 부정 감성

In [42]:
from nltk.stem import WordNetLemmatizer
from nltk.corpus import sentiwordnet as swn
from nltk import sent_tokenize, word_tokenize, pos_tag

def swn_polarity(text):
    # 감성 지수 초기화
    sentiment = 0.0
    tokens_count = 0

    lemmatizer = WordNetLemmatizer()
    raw_sentences = sent_tokenize(text)
    # 분해된 문장별로 단어 토큰 -> 품사 태깅 후에 SentiSynset 생성 -> 감성 지수 합산
    for raw_sentence in raw_sentences:
        # NTLK 기반의 품사 태깅 문장 추출
        tagged_sentence = pos_tag(word_tokenize(raw_sentence))
        for word , tag in tagged_sentence:

            # WordNet 기반 품사 태깅과 어근 추출
            wn_tag = penn_to_wn(tag)
            if wn_tag not in (wn.NOUN , wn.ADJ, wn.ADV):
                continue
            lemma = lemmatizer.lemmatize(word, pos=wn_tag)
            if not lemma:
                continue
            # 어근을 추출한 단어와 WordNet 기반 품사 태깅을 입력해 Synset 객체를 생성.
            synsets = wn.synsets(lemma , pos=wn_tag)
            if not synsets:
                continue
            # sentiwordnet의 감성 단어 분석으로 감성 synset 추출
            # 모든 단어에 대해 긍정 감성 지수는 +로 부정 감성 지수는 -로 합산해 감성 지수 계산.
            synset = synsets[0]
            swn_synset = swn.senti_synset(synset.name())
            sentiment += (swn_synset.pos_score() - swn_synset.neg_score())
            tokens_count += 1

    if not tokens_count:
        return 0

    # 총 score가 0 이상일 경우 긍정(Positive) 1, 그렇지 않을 경우 부정(Negative) 0 반환
    if sentiment >= 0 :
        return 1

    return 0


swn_polarity(text) 함수를 IMDB 감상평의 개별 문서에 적용해 긍정 및 부정 감성을 예측

In [43]:
review_df['preds'] = review_df['review'].apply( lambda x : swn_polarity(x) )
y_target = review_df['sentiment'].values
preds = review_df['preds'].values

SentiWordNet의 감성 분석 예측 성능 살펴보기

In [44]:
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score
from sklearn.metrics import recall_score, f1_score, roc_auc_score
import numpy as np

print(confusion_matrix( y_target, preds))
print("정확도:", np.round(accuracy_score(y_target , preds), 4))
print("정밀도:", np.round(precision_score(y_target , preds),4))
print("재현율:", np.round(recall_score(y_target, preds), 4))

[[7669 4831]
 [3635 8865]]
정확도: 0.6614
정밀도: 0.6473
재현율: 0.7092


### [VADER을 이용한 감성 분석]

VADER
- 또다른 Lexicon인 VADER Lexicon
- 소셜 미디어의 감성 분석 용도로 만들어진 룰 기반의 Lexicon


In [45]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer
senti_analyzer = SentimentIntensityAnalyzer()
senti_scores = senti_analyzer.polarity_scores(review_df['review'][0])
print(senti_scores)

{'neg': 0.13, 'neu': 0.743, 'pos': 0.127, 'compound': -0.7943}


매우 쉽게 감성 분석 수행할 수 있음
- SentimentIntensityAnalyzer 객체 생성 후 문서 별로 polarity_scores() 메서드를 호출해 감성 점수 구함
- 감성 점수가 특정 임계값 이상이면 긍정, 아니면 부정으로 판단

SentimentIntensityAnalyzer 객체의 polarity_scores() 메서드 : 딕셔너리 형태의 감성 점수를 반환
- 'neg' : 부정 감성 지수
- 'neu' : 중립 감성 지수
- 'pos' : 긍정 감성 지수
- compound는 neg, neu, pos score를 적절히 조합해 -1에서 1 사이의 감성 지수를 표현한 값

In [46]:
def vader_polarity(review,threshold=0.1):
    analyzer = SentimentIntensityAnalyzer()
    scores = analyzer.polarity_scores(review)

    # compound 값에 기반하여 threshold 입력값보다 크면 1, 그렇지 않으면 0을 반환
    agg_score = scores['compound']
    final_sentiment = 1 if agg_score >= threshold else 0
    return final_sentiment

# apply lambda 식을 이용하여 레코드별로 vader_polarity( )를 수행하고 결과를 'vader_preds'에 저장
review_df['vader_preds'] = review_df['review'].apply( lambda x : vader_polarity(x, 0.1) )
y_target = review_df['sentiment'].values
vader_preds = review_df['vader_preds'].values

print(confusion_matrix( y_target, vader_preds))
print("정확도:", np.round(accuracy_score(y_target , vader_preds),4))
print("정밀도:", np.round(precision_score(y_target , vader_preds),4))
print("재현율:", np.round(recall_score(y_target, vader_preds),4))

[[ 6747  5753]
 [ 1858 10642]]
정확도: 0.6956
정밀도: 0.6491
재현율: 0.8514


-> 정확도가 SentiWordNet 보다 향상됨